app/shared/design-system/components/primitives/auto-input.tsx

```diff
@@ -0,0 +1,205 @@
+export class AutoInput extends EventTarget {
+	// PRIVATE
+	#selectTarget = null;
+	#activeTarget = null;
+	#activeEvents = ['keydown', 'focusin', 'focusout', 'pointerover', 'pointerout', 'pointerup', 'pointerdown'];
+	#passiveEvents = ['pointermove'];
+	#resizeCursor = 'ns-resize';
+	#pointerCursor = 'pointer';
+	#pointerCapture = false;
+	#ownerDocument = globalThis.document;
+	#canvasContext = globalThis.document?.createElement('canvas').getContext('2d');
+	#typeOfValue(value) {
+		if (/^([-+]?\d*\.?\d+)$/.test(value)) return 'number';
+		if (/^(%|cm|mm|in|px|pt|pc|em|ex|ch|rem|vw|vh|vmin|vmax|vb|vi|svw|svh|lvw|lvh|dvw|dvh|lh|rlh)$/.test(value)) return 'unit';
+		if (/^(\w+)$/.test(value)) return 'keyword';
+		if (/^(\w+\()/.test(value)) return 'function';
+		return '';
+	}
+	#measureEvent(value, event, target) {
+		const {fontSize, fontFamily, fontWeight, textAlign, paddingLeft, paddingRight} = getComputedStyle(target);
+		this.#canvasContext.font = `${fontWeight} ${fontSize} ${fontFamily}`;
+		const eventOffset = event.offsetX;
+		let charsOffset = 0;
+		let pixelPadding = 0
+		switch (textAlign) {
+			case 'left': {
+				pixelPadding += parseFloat(paddingLeft);
+				break;
+			}
+			case 'right': {
+				pixelPadding += target.clientWidth - this.#canvasContext.measureText(value).width - parseFloat(paddingRight);
+				break;
+			}
+		}
+		let pixelOffset = pixelPadding;
+		let valueOffset = -1;
+		const valueTokens = value.split(/([*/,()\s]?)([-+]?\d*\.?\d+)|([*/,()\s]+)/).filter(value => value).map(value => {
+			return {tokenType: this.#typeOfValue(value), tokenValue: value}
+		});
+		for (const [tokenIndex, tokenObject] of valueTokens.entries()) {
+			const {tokenValue} = tokenObject;
+			const tokenOffset = this.#canvasContext.measureText(tokenValue).width;
+			if ((eventOffset >= pixelOffset && eventOffset <= pixelOffset + tokenOffset)) {
+				valueOffset = tokenIndex;
+				break;
+			}
+			charsOffset += tokenValue.length;
+			pixelOffset += tokenOffset;
+		}
+		return {valueOffset, valueTokens, charsOffset, pixelOffset, pixelPadding};
+	}
+	// PUBLIC
+	handleEvent(event) {
+		const eventTarget = event.target;
+		const currentTarget = this.#activeTarget || eventTarget;
+		if (currentTarget?.nodeName !== 'INPUT' || currentTarget?.type !== 'text' || currentTarget?.readOnly !== false) return;
+		switch (event.type) {
+			case 'pointerover': case 'focusin': {
+				if (this.#pointerCapture === false) this.#pointerCapture = !this.#passiveEvents.forEach(eventName => this.#ownerDocument.addEventListener(eventName, this, false));
```

Comment:
`!this.#passiveEvents.forEach`

am I missing something or is it basically `![]` and will always be false?

Reply:
right, but why would you write it this way? it looks like its made so that someone can't read it, extra cryptic lol